fix: add telemetry reporting

dgolovin opened the pull request:
Fix #43.

slemeur commented on src/extension.ts:
```diff
@@ -352,13 +359,19 @@ export async function activate(context: extensionApi.ExtensionContext): Promise<
+
   });
 
   const SignOutCommand = extensionApi.commands.registerCommand('redhat.authentication.signout', async () => {
```
There should be an event on signout too

slemeur commented on src/extension.ts:
```diff
@@ -352,13 +359,19 @@ export async function activate(context: extensionApi.ExtensionContext): Promise<
       removeSession(currentSession.id); // if at least one fail, remove session
     }
+
+    TelemetryLogger.logUsage('redhat.authentication.signin', telemetryData);
```
For consistency and accuracy, I think the name of the event should be using the name of the extension:
`redhat.redhat-authentication`

jeffmaury replied:
The extension id is added by the Podman Desktop API so not required here. signin should be enough

dgolovin replied:
Updated logUsage call to use short event names like `signin`, `signout` and etc

dgolovin commented on src/podman-cli.ts:
```diff
@@ -89,6 +92,7 @@ export async function runRpmInstallSubscriptionManager(): Promise<number | undef
   } catch (err) {
     const exitCode = (err as extensionApi.RunError).exitCode;
     console.error(`Subscription manager installation returned exit code: ${exitCode}`);
+    TelemetryLogger.logError('SubscriptionManagerInstallationError', {error: String(err)});
```
```suggestion
    TelemetryLogger.logError('subscriptionManagerInstallationError', {error: String(err)});
```

dgolovin commented on src/extension.ts:
```diff
@@ -319,7 +323,10 @@ export async function activate(context: extensionApi.ExtensionContext): Promise<
       .then(() => false)
-      .catch(() => true); // required to force Promise.all() call keep waiting for all not failed calls
+      .catch((error) => {
+        telemetryData['registryAccessConfigurationError'] = String(error);
```
```suggestion
```
not required, because it is going to be recorded in podman-cli.ts

dgolovin commented on src/extension.ts:
```diff
@@ -352,20 +359,27 @@ export async function activate(context: extensionApi.ExtensionContext): Promise<
       .then(() => false)
-      .catch(() => true); // required to force Promise.all() call keep waiting for all not failed calls
+      .catch((error) => {
+        telemetryData['vmSubscriptionActivationError'] = String(error);
```
```suggestion
```

not required, because it is going to be recorded in podman-cli.ts

dgolovin commented on src/extension.ts:
```diff
@@ -302,6 +304,8 @@ export async function activate(context: extensionApi.ExtensionContext): Promise<
   context.subscriptions.push(providerDisposable);
 
   const SignInCommand = extensionApi.commands.registerCommand('redhat.authentication.signin', async () => {
+    const telemetryData = {};
```
```suggestion
    const telemetryData = {
      successful: true,
    };
```

dgolovin commented on src/extension.ts:
```diff
@@ -352,20 +359,27 @@ export async function activate(context: extensionApi.ExtensionContext): Promise<
 
     if (failed.some(fail => fail)) {
       removeSession(currentSession.id); // if at least one fail, remove session
     }
+
```
```suggestion
      telemetryData.successful = false;
    }

```

benoitf commented on src/extension.ts:
```diff
@@ -352,20 +360,27 @@ export async function activate(context: extensionApi.ExtensionContext): Promise<
+      telemetryData.successful = false;
     }
+
+    TelemetryLogger.logUsage('signin', telemetryData);
```
I think we're missing lot of data in telemetryData

here it's only successful yes/no

I would expect the time it has taken, in case of failure, which step succeed/failed.


did we have to restart and how many restart

at which step the user stopped the flow

for example it's different if user aborted if no podman machine was running vs error in doing the task
did we have a PodmanMachineRunning ?

was isSubscriptionManagerInstalled already installed ?

Was it registryAccess part, vmActivation part

for each step, the time it took as well.

Then about the session, probably need some data (about the user being authenticated
I don't know what data we have but maybe number about some fields of the user (counter)

dgolovin replied:
@benoitf @slemeur would be possible to fix all of that as follow up issues? I have added part of it and working on tests right now, but it will take some time to implement.

vrothberg replied:
I think that's fair.  We had a meeting on the telemetry issue (https://github.com/redhat-developer/podman-desktop-redhat-account-ext/issues/43#issuecomment-1997959842) where these details did not come up.  So I guess we under-specified the requirements initially.

Merging the PR and trying to get as far as possible until the deadline is OK for me.

dgolovin commented:
@benoitf 

> I think we're missing lot of data in telemetryData
> I would expect the time it has taken, in case of failure, which step succeed/failed.

> did we have to restart and how many restart

It is always one restart if needed at all, I can send start and stop command outcome.

> at which step the user stopped the flow

User cannot stop the flow right now, it keeps going until it is finished with failure or success.

> for example it's different if user aborted if no podman machine was running vs error in doing the task
did we have a PodmanMachineRunning ?

I have added error messages to telemetry for registry configuration and subscription activation

> was isSubscriptionManagerInstalled already installed ?

Follow up issue #113 created

> Was it registryAccess part, vmActivation part

added field with error message for each part

> for each step, the time it took as well.

Follow up issue #114 created

> Then about the session, probably need some data (about the user being authenticated
> I don't know what data we have but maybe number about some fields of the user (counter)

I need clarification for this. What king of tracking we want to do here. Number of logins for specific user? Session duration?
